From e83a2c56154aa006e4d990d835516955f9480135 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Fri, 15 Sep 2023 15:26:50 +0200
Subject: [PATCH 0375/1695] drivers: stm32mp_provisioning: support low-power
 sequence

In some low-power modes such as standby, the BSEC content is lost
after a BSEC Warm reset and the provisioning values must be restored
in shadow register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 30a2b20449ae1b7c520c1720fc2171bbf0e81485 drivers: stm32mp_provisioning: support low-power sequence]
Change-Id: I596056835a178032dbeaae53b9f37b13b04fce8d
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/328011
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/stm32mp_provisioning.c | 27 +++++++++++++++++++++------
 1 file changed, 21 insertions(+), 6 deletions(-)

diff --git a/core/drivers/stm32mp_provisioning.c b/core/drivers/stm32mp_provisioning.c
index b51917d87..6f3460bd5 100644
--- a/core/drivers/stm32mp_provisioning.c
+++ b/core/drivers/stm32mp_provisioning.c
@@ -12,6 +12,7 @@
 #include <kernel/dt.h>
 #include <kernel/dt_driver.h>
 #include <kernel/panic.h>
+#include <kernel/pm.h>
 #include <libfdt.h>
 #include <stdbool.h>
 #include <stdlib.h>
@@ -109,8 +110,7 @@ static TEE_Result provisioning_subnode(const void *fdt, int node)
 	return res;
 }
 
-#ifdef CFG_DT
-static TEE_Result provisioning_probe(void)
+static TEE_Result provisioning_init(void)
 {
 	const void *fdt = get_embedded_dt();
 	int node = -1;
@@ -124,11 +124,26 @@ static TEE_Result provisioning_probe(void)
 
 	return provisioning_subnode(fdt, node);
 }
-#else
-static TEE_Result provisioning_probe(void)
+
+static TEE_Result provisioning_pm(enum pm_op op, unsigned int pm_hint __unused,
+				  const struct pm_callback_handle *hdl __unused)
 {
+	if (op == PM_OP_RESUME)
+		return provisioning_init();
+
 	return TEE_SUCCESS;
 }
-#endif
 
-early_init(provisioning_probe);
+static TEE_Result provisioning_probe(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = provisioning_init();
+
+	if (res == TEE_SUCCESS && IS_ENABLED(CFG_PM))
+		register_pm_driver_cb(provisioning_pm, NULL, "provisioning");
+
+	return res;
+}
+
+early_init_late(provisioning_probe);
-- 
2.47.2

